cmd/errtrace/toolexec.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"braces.dev/errtrace"
+)
+
+func (cmd *mainCmd) handleToolExec(args []string) (exitCode int, handled bool) {
+	// In toolexec mode, we're passed the original command + arguments.
+	if len(args) == 0 {
+		return -1, false
+	}
+
+	for _, arg := range args {
+		if arg == "-V=full" {
+			// compile is run first with "-V=full" to get a version number
+			// for caching build IDs.
+			// No TOOLEXEC_IMPORTPATH is set in this case.
+			return cmd.toolExecVersion(args), true
+		}
+	}
+
+	if cmd.Getenv == nil {
+		cmd.Getenv = os.Getenv
+	}
+	// When "-toolexec" is used, the go cmd sets the package being compiled in the env.
+	if pkg := cmd.Getenv("TOOLEXEC_IMPORTPATH"); pkg != "" {
+		return cmd.toolExecRewrite(pkg, args), true
+	}
+
+	return -1, false
+}
+
+func (cmd *mainCmd) toolExecVersion(args []string) int {
+	tool := exec.Command(args[0], args[1:]...)
+	var stdout bytes.Buffer
+	tool.Stdout = &stdout
+	tool.Stderr = cmd.Stderr
+	tool.Env = os.Environ()
+	if err := tool.Run(); err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			return exitErr.ExitCode()
+		}
+
+		fmt.Fprintf(cmd.Stderr, "%v failed: %v", args[0], err)
+		return 1
+	}
+
+	// TODO: This version number should change whenever the rewriting changes.
+	fmt.Fprintf(cmd.Stdout, "%s-errtrace0\n", strings.TrimSpace(stdout.String()))
+	return 0
+}
+
+func (cmd *mainCmd) toolExecRewrite(pkg string, args []string) (exitCode int) {
+	// We only need to modify the arguments for "compile" calls which work with .go files.
+	if !isCompile(args[0]) {
+		return cmd.runOriginal(args)
+	}
+
+	// We only modify files that import errtrace, and have "errtrace.ToolExecInstrument"
+	// However, that requires parsing files to determine, which we want to avoid for stdlib
+	// so use a heuristic to detect stdlib packages -- whether the name contains "."".
+	if !strings.Contains(pkg, ".") {
+		return cmd.runOriginal(args)
+	}
+
+	exitCode, err := cmd.rewriteCompile(pkg, args)
+	if err != nil {
+		cmd.log.Print(err)
+		return 1
+	}
+
+	return exitCode
+}
+
+func (cmd *mainCmd) rewriteCompile(pkg string, args []string) (exitCode int, _ error) {
+	parsed := make(map[string]file)
+	var canRewrite, needRewrite bool
+	for _, arg := range args {
+		if !isGoFile(arg) {
+			continue
+		}
+
+		contents, err := os.ReadFile(arg)
+		if err != nil {
+			return -1, errtrace.Wrap(err)
```

Comment:
It's mostly as a safety, just in case a change accidentally causes the exitCode to be used even in the error case, I wanted to avoid the command exiting successfully in error cases.